File: team7026/src/main/java/com/jdroids/team7026/ftc2017/MecanumDriveHelper.java
package com.jdroids.team7026.ftc2017;

import com.jdroids.team7026.lib.util.transformations.DriveSignal;

public class MecanumDriveHelper {
    public DriveSignal angleDrive(double angle, double power, boolean breakMode) {//Angle in radians
        return new DriveSignal(power * Math.cos(angle), -power * Math.sin(angle),
                power * Math.sin(angle), -power * Math.cos(angle), breakMode);
    }
    public DriveSignal angleDrive(double angle, double power) {
        return angleDrive(angle, power, true);
    }

    public DriveSignal arcadeDrive(double forwardsVelocity, double sidewaysVelocity,
                                   double angularVelocity, boolean fieldOrientedDriveEnabled,
                                   double currentRadians, double offset, boolean breakMode) {
        double power = Math.hypot(sidewaysVelocity, forwardsVelocity);

        double angle;
        if(fieldOrientedDriveEnabled) {
            angle = (offset + Math.atan2(forwardsVelocity, sidewaysVelocity) - currentRadians) - Math.PI / 4;
        }
        else {
            angle = Math.atan2(forwardsVelocity, sidewaysVelocity) - Math.PI / 4;
        }

        return new DriveSignal(
                power * Math.cos(angle) + angularVelocity,
                -power * Math.sin(angle) - angularVelocity,
                power * Math.sin(angle) + angularVelocity,
                -power * Math.cos(angle) - angularVelocity);
    }

    public DriveSignal arcadeDrive(double forwardsVelocity, double sidewaysVelocity,
                                   double angularVelocity, boolean fieldOrientedDriveEnabled,
                                   double currentRadians, double offset) {
        return arcadeDrive(forwardsVelocity, sidewaysVelocity, angularVelocity,
                fieldOrientedDriveEnabled, currentRadians, offset);
    }

    public DriveSignal arcadeDrive(double forwardsVelocity, double sidewaysVelocity,
                                   double angularVelocity, boolean fieldOrientedDriveEnabled,
                                   double currentRadians) {
        return arcadeDrive(forwardsVelocity, sidewaysVelocity, angularVelocity,
                fieldOrientedDriveEnabled, currentRadians, 0, true);
    }

    public DriveSignal arcadeDrive(double forwardsVelocity, double sidewaysVelocity,
                                   double angularVelocity) {
        return arcadeDrive(forwardsVelocity, sidewaysVelocity, angularVelocity,
                false, 0, 0);
    }
}

File: team7026/src/main/java/com/jdroids/team7026/ftc2017/MecanumDriveHelper.kt
package com.jdroids.team7026.ftc2017

import com.jdroids.team7026.lib.util.DriveSignal
import com.jdroids.team7026.ftc2017.Robot.getOrientationInRadians

object MecanumDriveHelper {
    fun angleDrive(angle: Double, power: Double, brakeMode: Boolean = false): DriveSignal {
        //Angle is in radians
        return DriveSignal(power * Math.cos(angle), -power * Math.sin(angle),
                power * Math.sin(angle), -power * Math.cos(angle), brakeMode)
    }

    fun arcadeDrive(forwardsVelocity: Double, sidewaysVelocity: Double,
                    angularVelocity: Double, fieldOrientedDriveEnabled: Boolean = false,
                    offset: Double = 0.0, brakeMode: Boolean = false): DriveSignal {
        val power = Math.hypot(sidewaysVelocity, forwardsVelocity)


        val currentRadians= if(fieldOrientedDriveEnabled) {
            Robot.imu!!.getOrientationInRadians().firstAngle.toDouble()
        }
        else {0.0}

        val angle = offset + Math.atan2(forwardsVelocity, sidewaysVelocity) - currentRadians - Math.PI / 4



        return DriveSignal(
                power * Math.cos(angle) - angularVelocity,
                -power * Math.sin(angle) - angularVelocity,
                power * Math.sin(angle) - angularVelocity,
                -power * Math.cos(angle) - angularVelocity)
    }
}